File: Android/WaterContainer/java/java/com/guildon/watercontainer/WaterContainerView.java
package com.guildon.watercontainer;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.View;

/**
 * Created by Chris on 21/Apr/2018.
 * <p>
 * Java version of the View which displays the wall configuration,
 * and the water it stores, graphically.
 */
public class WaterContainerView extends View {

	/**
	 * Paint object used to draw a grid
	 */
	private final Paint linePaint = new Paint();

	/**
	 * Paint object used to draw a wall
	 */
	private final Paint wallPaint = new Paint();

	/**
	 * Paint object used to draw water
	 */
	private final Paint waterPaint = new Paint();

	/**
	 * Layout of the walls and water
	 */
	private char[] layout;

	/**
	 * Width of the layout
	 */
	private int layoutWidth = 0;

	/**
	 * Height of the layout
	 */
	private int layoutHeight = 0;

	/**
	 * Width and height of a "cell" in the layout
	 */
	private final int cellSize = 20;


	public WaterContainerView(Context context, AttributeSet set) {
		super(context, set);

		wallPaint.setColor(Color.GRAY);
		waterPaint.setColor(Color.CYAN);
		linePaint.setColor(Color.BLACK);

		linePaint.setStrokeWidth(2);
	}

	/**
	 * Set the layout to show in this View
	 *
	 * @param mLayout : The layout to show in the view
	 * @param mWidth  : Width of the layout
	 * @param mHeight : Maximum height of the layout
	 */
	public void setLayout(char[] mLayout, int mWidth, int mHeight) {
		layout = mLayout;

		layoutWidth = mWidth;
		layoutHeight = mHeight;
	}


	@Override
	protected void onDraw(Canvas canvas) {

		if (layout == null)
			return;

		// Pass through the layout, rendering each cell
		for (int i = 0; i < layoutHeight; i++) {
			for (int j = 0; j < layoutWidth; j++) {
				switch (layout[((layoutHeight - 1 - i) * layoutWidth) + j]) {
					case '#':
						canvas.drawRect(j * cellSize, i * cellSize, (j + 1) * cellSize, (i + 1) * cellSize, wallPaint);
						break;
					case 'w':
						canvas.drawRect(j * cellSize, i * cellSize, (j + 1) * cellSize, (i + 1) * cellSize, waterPaint);
						break;
				}
			}
		}

		// Draw a grid to show each cell
		for (int i = 1; i < layoutHeight; i++)
			canvas.drawLine(0.0f, i * cellSize, layoutWidth * cellSize, i * cellSize, linePaint);

		for (int i = 0; i < layoutWidth; i++)
			canvas.drawLine(i * cellSize, 0.0f, i * cellSize, layoutHeight * cellSize, linePaint);
	}
}

File: Android/WaterContainer/kotlin/java/com/guildon/watercontainer/WaterContainerView.kt
package com.guildon.watercontainer

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import android.view.View

/**
 * Created by Chris Allen on 14/Sep/2018.
 *
 * Kotlin version of the View which displays the wall configuration,
 * and the water it stores, graphically.
 */
class WaterContainerView(context: Context, attrs: AttributeSet) : View(context, attrs) {

	/**
	 * Paint object used to draw a grid
	 */
	private val linePaint = Paint()

	/**
	 * Paint object used to draw a wall
	 */
	private val wallPaint = Paint()

	/**
	 * Paint object used to draw water
	 */
	private val waterPaint = Paint()

	/**
	 * Width of the layout
	 */
	private var layoutWidth: Int = 0

	/**
	 * Height of the layout
	 */
	private var layoutHeight: Int = 0

	/**
	 * Layout of the walls and water
	 */
	private lateinit var layout: Array<Char>

	/**
	 * Width and height of a "cell" in the layout
	 */
	private val cellSize = 20.0f

	init {
		linePaint.color = Color.BLACK
		linePaint.strokeWidth = 2.0f

		wallPaint.color = Color.GRAY
		waterPaint.color = Color.CYAN
	}

	/**
	 * Set the layout to show in this View
	 *
	 * @param mLayout : The layout to show in the view
	 * @param mWidth : Width of the layout
	 * @param mHeight : Maximum height of the layout
	 */
	fun setLayout(mLayout: Array<Char>, mWidth: Int, mHeight: Int) {
		layout = mLayout

		layoutWidth = mWidth
		layoutHeight = mHeight
	}


	override fun onDraw(canvas: Canvas) {

		// Pass through the layout, rendering each cell
		for (i in 0 until layoutHeight) {
			for (j in 0 until layoutWidth) {
				when (layout[((layoutHeight - 1 - i) * layoutWidth) + j]) {
					'#' -> canvas.drawRect(j * cellSize, i * cellSize, (j + 1) * cellSize, (i + 1) * cellSize, wallPaint)
					'w' -> canvas.drawRect(j * cellSize, i * cellSize, (j + 1) * cellSize, (i + 1) * cellSize, waterPaint)
				}
			}
		}

		// Draw a grid to show each cell
		for (i in 1 until layoutHeight)
			canvas.drawLine(0.0f, i * cellSize, layoutWidth * cellSize, i * cellSize, linePaint)

		for (i in 0 until layoutWidth)
			canvas.drawLine(i * cellSize, 0.0f, i * cellSize, layoutHeight * cellSize, linePaint)
	}


}